### ISSUE TITLE
Evicted pods cannot be rescheduled

### ISSUE THREAD

**Issue by matthias50** [cybozu-go/topolvm#41]
**Describe the bug**
Once a pod is scheduled to a given node and the LVM created and the PVC is bound, the pod cannot be evicted and scheduled to another node.

**Environments**
- Version: Any
- OS: Any

**To Reproduce**
1. Deploy a stateful set with a volume claim template in the namespace "testnamespace" Example yaml is below
```(yaml)
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: test-lvm
  namespace: testnamespace
spec:
  serviceName: "test-lvm"
  replicas: 5
  podManagementPolicy: Parallel
  selector:
    matchLabels:
      app: test-lvm
  template:
    metadata:
      labels:
        app: test-lvm
    spec:
      containers:
        - name: test-lvm-container
          image: busybox
          command:
            - sleep
            - "3600"
          volumeMounts:
            - mountPath: "/mnt/storage"
              name: test-lvm-pvc

  volumeClaimTemplates:
  - metadata:
      name: test-lvm-pvc
    spec:
      accessModes: [ "ReadWriteOnce" ]
      storageClassName: topolvm-provisioner
      resources:
        requests:
          storage: 1Gi
```
2. 

```
$ kubectl drain --ignore-daemonsets $(kubectl -n testnamespace get po -o jsonpath='{.spec.nodeName}' test-lvm-0)
...
evicting pod "test-lvm-4"
evicting pod "test-lvm-0"
evicting pod "test-lvm-2"
node/redacted-name
```
3.  Pod(s) are stuck in Pending state.
```
$ kubectl -n testnamespace get po
NAME         READY   STATUS    RESTARTS   AGE
test-lvm-0   0/1     Pending   0          93s
test-lvm-1   1/1     Running   0          9m3s
test-lvm-2   0/1     Pending   0          93s
test-lvm-3   1/1     Running   0          9m3s
test-lvm-4   0/1     Pending   0          93s
```

**Expected behavior**

I expect the PVC/PV to automatically be deleted from the node where the pod was running and a new PVC/PV is created on the node where the pod gets scheduled. Since this storage is ephemeral, once the pod leaves the node it should immediately be cleaned up.

**Additional context**

I was able to manually recover the stateful set by deleting the PVC and then the pod.

```
$ kubectl -n testnamespace delete pvc test-lvm-pvc-test-lvm-0
$ kubectl -n testnamespace delete po test-lvm-0
$ kubectl -n testnamespace get po
NAME         READY   STATUS    RESTARTS   AGE
test-lvm-0   1/1     Running   0          13s
test-lvm-1   1/1     Running   0          11m
test-lvm-2   0/1     Pending   0          3m32s
test-lvm-3   1/1     Running   0          11m
test-lvm-4   0/1     Pending   0          3m32s
```

**Comment by ymmt2005** [comment]
@matthias50 
I think deleting TopoLVM is not responsible to delete PVC when Pod is deleted
because TopoLVM did *not* create the PVC.  It is Kubernetes that created PVC
from the template.

**Comment by ymmt2005** [comment]
The current implementation allows a new Pod is rescheduled to the same
node and therefore can re-use the PVC/PV when, for example, a container
image is updated.  I think this is a preferred behavior.

We have a plan to automatically remove PVC/PV when the node where
the volume exists is deleted from Kubernetes.

**Comment by ymmt2005** [comment]
Thank you for the scenario.

We will soon investigate the problem here further in https://github.com/cybozu-go/neco/issues/329 probably in a month or so.
Will update.

**Comment by matthias50** [comment]
@ymmt2005, looks like there has been a lot of progress on this. Thanks for taking this use case in to account. One concern I have from looking through the commits is that they seem to be focused just on cleaning up after a node is completely gone. 

Here is a scenario to consider along with how I would like it to work (which is how it does work when topolvm isn't in the picture)

Pods A and B are running on node X. Both Pod A and Pod B are using LVM PVs.
Pod A is evicted, but pod B can't be evicted because of Pod disruption budget.
Pod A is able to be immediately rescheduled to node Y even though node X is still up.

Is this how it will work, or would node X have to die before any of the pods which were there could be rescheduled to a different node?

**Comment by ymmt2005** [comment]
@matthias50 
Yes, our improvements so far are cleanups after node removal, which itself is nice.
As long as a node survives, volumes on the node are retained.

To automatically remove volumes of evicted pods, another work is needed.
Please let me know if this fulfills your scenario or not:

1. A node is marked unschedulable by `kubectl drain` or `kubectl cordon`.
2. Pods on the node are evicted.
3. PVCs/PVs once bound for the pods are automatically removed if:
    - they are on the unschedulable node, and
    - they are marked as _ephemeral_ by, say, an annotation.

Please note that we want to keep PVCs/PVs for quick node maintenance (reboot).

**Comment by matthias50** [comment]
That's close to what I had in mind, though I'm not sure why the annotation would be needed.

> Please note that we want to keep normal PVCs/PVs for quick node maintenance (reboot).

Does this mean that you would expect that if a node reboots that all the pods would come back up with their PVCs/PVs after the reboot?

I was thinking about looking at this more from the perspective of the Pod. I was thinking more like setting up a watch for pods stuck in the Pending state for some period of time (say 10 seconds).

To me, this is the indication that something  may have gone awry. So if a Pod is in the Pending state and it has a PVC provisioned by the topolvm provisioner then I would want to delete the PV so that the pod could be rescheduled. Not sure how to convince the PVC that it is "unbound" at that point though. And I think this would break your reboot use case.

Honestly, while a PVC being trapped on a node which has been cordoned is the most common case, there  are others such as this (which is a race condition):

1. Delete pod A (with a topolvm PVC) from node X
2. Pod B is scheduled to node X
3. Pod A is unable to be scheduled to node X because of some other resource problem (say insufficient CPU or Memory).

This will also deadlock requiring manual intervention. Scenarios such as this lead me to believe that it is best to always view the LVMs as ephemeral and delete them and the PV whenever the Pod with the PVC leaves the node for any reason).

Regardless of the technical details of how this is accomplished, what is key for me is to not have Pods stuck in the Pending state because of LVMs being on inaccessible nodes *ever*.  We run clusters with hundreds of r4.8xlarge instances and a heterogenous workload which we, the platform providers, don't control. So I'm willing to treat LVMs as purely ephemeral to achieve reliable scheduling and avoid calls in the middle of the night :) Since the LVMs are on a single node which could have a hardware failure, applications running on the platform can't depend on the data being stored on a node being durable anyway.

As an interesting aside, we don't support reboot of nodes *at all* if one of our nodes reboots, we detect the condition and terminate it during boot up. Our stance is that nodes are immutable disposable infrastructure, so we never update them or reboot them. This helps us avoid dealing with special cases (initial boot, vs. reboot etc.). I'm not trying to convince you that this is how to run your cluster, just letting you know where we are coming from.

**Comment by ymmt2005** [comment]
Thank you for clarification.  We understood your use case.

We use TopoLVM for our on-premise data centers.
Since nodes are physical servers, we often reboot servers to update firmware and OS.

For your use case, I'd like to suggest two solutions:

1. Watch and delete pending TopoLVM pods externally.
2. Enhance `topolvm-controller` to watch pending TopoLVM pods and delete them.
    This feature is off by default.  To enable, pass `--clean-pending-pvc=<duration>` command-line flag.

If you are willing to create a patch for option 2, we welcome and review your pull request.
If not, we may implement it in the future after other prioritized features are implemented.

**Comment by matthias50** [comment]
Those options make sense. I should have some cycles in a few weeks to take a look at this. I'll see if I can make option 2 work and fall back to option 1.

**Comment by ymmt2005** [comment]
Resolved with #75.
